File: 380/380_insert_delete_get_random_o1.py
import random


class RandomizedSet:

    def __init__(self):
        self.map = {}
        self.arr = []

    def insert(self, val: int) -> bool:
        if val in self.map:
            return False
        
        self.arr.append(val)
        self.map[val] = len(self.arr) - 1
        return True
        

    def remove(self, val: int) -> bool:
        if val not in self.map:
            return False
        
        last_element = self.arr[-1]
        idx_of_val_to_remove = self.map[val]

        self.arr[idx_of_val_to_remove] = last_element
        self.map[last_element] = idx_of_val_to_remove

        self.arr.pop()
        del self.map[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.arr)


if __name__ == "__main__":
    obj = RandomizedSet()

    commands = ["RandomizedSet", "remove", "remove", "insert", "getRandom", "remove", "insert"]
    arguments = [[], [0], [0], [0], [], [0], [0]]

    # Simulate input
    for i, command in enumerate(commands):
        if command == "RandomizedSet":
            print("null")
        elif command == "insert":
            result = obj.insert(arguments[i][0])
            print(result)
        elif command == "remove":
            result = obj.remove(arguments[i][0])
            print(result)
        elif command == "getRandom":
            result = obj.getRandom()
            print(result)

File: 380/380_insert_delete_get_random_o1.cpp
# include <iostream>
# include <vector>
# include <unordered_map>
# include <ctime>
using namespace std;

class RandomizedSet {
private:
    unordered_map<int, int> map;
    vector<int> arr;

public:
    RandomizedSet() {
        srand(time(0));
    }

    bool insert(int val) {
        if (map.find(val) != map.end()) {
            return false;
        }
        arr.push_back(val);
        map[val] = arr.size() - 1;
        return true;
    }

    bool remove(int val) {
        if (map.find(val) == map.end()) {
            return false;
        }
        int last = arr.back();
        int idx = map[val];

        arr[idx] = last;
        map[last] = idx;

        arr.pop_back();
        map.erase(val);
        return true;
    }

    int getRandom() {
        int randIndex = rand() % arr.size();
        return arr[randIndex];
    }
};

int main() {
    vector<string> commands = {"RandomizedSet", "remove", "remove", "insert", "getRandom", "remove", "insert"};
    vector<vector<int>> arguments = {{}, {0}, {0}, {0}, {}, {0}, {0}};

    RandomizedSet* obj = nullptr;

    for (int i = 0; i < commands.size(); i++) {
        if (commands[i] == "RandomizedSet") {
            obj = new RandomizedSet();
            cout << "null" << endl;  // Constructor does not return a value
        } else if (commands[i] == "insert") {
            bool result = obj->insert(arguments[i][0]);
            cout << (result ? "true" : "false") << endl;
        } else if (commands[i] == "remove") {
            bool result = obj->remove(arguments[i][0]);
            cout << (result ? "true" : "false") << endl;
        } else if (commands[i] == "getRandom") {
            int result = obj->getRandom();
            cout << result << endl;
        }
    }

    delete obj;  // Free the allocated memory

    return 0;
}